### DISCUSSION TITLE
(D::Larger)::Smaller != D ?

### DISCUSSION THREAD

**Discussion by goertzenator** [rust-ndarray/ndarray#1461] [Q&A] [answered]
I am writing an ndarray function that is generic in dimension (`D: Dimension`) and am ending up with the following error when I try to Zip some things together:

```
    = note: expected associated type `<<<D as ndarray::Dimension>::Larger as ndarray::Dimension>::Smaller as ndarray::Dimension>::Smaller`
               found associated type `<D as ndarray::Dimension>::Smaller`
```

Intuitively these are the same dimension types, but I see how the Dimension trait doesn't enforce `(D::Larger)::Smaller` to be the same as `D`.  Any tips on establishing this constraint?

In the meantime I'm going to see if I can avoid using `Larger` to begin with.

**Comment by akern40** [comment]
Any chance you can provide a minimal example that produces this error? I'd be happy to look into it, but am honestly not sure how to trigger it!

**Reply by goertzenator** [reply]
Here is an example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9729023ceaf478356d661899d80ab5b9

@jturner314 explains why it can't work, and I have found a work around for my application.  However I am still mentioning this because I saw something about a redesign of the Dimension system.  Perhaps a different design could make this work?

**Comment by jturner314** [comment, marked as answer]
Unfortunately, they aren't always the same type. In the latest version of `ndarray`, constant-ndim types only go up to 6 dimensions, so [`Ix6::Larger = IxDyn`](https://docs.rs/ndarray/0.16.1/ndarray/trait.Dimension.html#associatedtype.Larger-7). Since [`IxDyn::Smaller = IxDyn`](https://docs.rs/ndarray/0.16.1/ndarray/trait.Dimension.html#associatedtype.Smaller-8), `<Ix6::Larger as Dimension>::Smaller = IxDyn::Smaller = IxDyn`.
